### PR TITLE
Add correct handling of server error responses in OTLP/HTTP exporter

### PR DESCRIPTION
**Pull request by tigrannajaryan**
This implements the response behavior according to the spec:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/otlp.md#response-1

### CODE REVIEW

File: exporter/otlphttpexporter/otlp.go [hunk 1]
```diff
@@ -109,15 +122,78 @@ func (e *exporterImp) export(ctx context.Context, url string, request []byte) er
...
+
+	// Report the failure as permanent if the server thinks the request is malformed.
+	if resp.StatusCode == http.StatusBadRequest {
+		return consumererror.Permanent(formattedErr)
+	}
+
+	return formattedErr
+}
+
+func readResponse(resp *http.Response) (*status.Status, error) {
```
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:176]: It doesn't seem that you do anything with the returned error, you can simply return just the status.

File: exporter/otlphttpexporter/otlp.go [hunk 2]
```diff
@@ -109,15 +122,78 @@ func (e *exporterImp) export(ctx context.Context, url string, request []byte) er
...
+			}
+		}
+		// Indicate to our caller to pause for the specified number of seconds.
+		return exporterhelper.NewThrottleRetry(formattedErr, time.Duration(retryAfter)*time.Second)
+	}
+
+	// Report the failure as permanent if the server thinks the request is malformed.
+	if resp.StatusCode == http.StatusBadRequest {
+		return consumererror.Permanent(formattedErr)
+	}
```
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:171]: You probably need a TODO for other cases where you can retry?

File: exporter/otlphttpexporter/otlp.go [hunk 3]
```diff
@@ -109,15 +122,78 @@ func (e *exporterImp) export(ctx context.Context, url string, request []byte) er
...
+	// Format the error message. Use the status if it is present in the response.
+	var formattedErr error
+	if err == nil && respStatus != nil {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d, Message=%s, Details=%v",
+			resp.StatusCode, respStatus.Message, respStatus.Details)
+	} else {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d",
+			resp.StatusCode)
```
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:150]: Should we retry in this case?

Comment (tigrannajaryan) [reply exporter/otlphttpexporter/otlp.go:150]: According to the spec we retry in all cases except http.StatusBadRequest. The presence or absence of the detailed error message does not influence that.

File: exporter/otlphttpexporter/otlp.go (hunk 2 above)
Comment (tigrannajaryan) [reply exporter/otlphttpexporter/otlp.go:171]: All error cases other than http.StatusBadRequest are retryable. That's we return non-permanent error by default. I believe that should result in a retry by caller (retrySender), right?
I will add a comment to explain.

File: exporter/otlphttpexporter/otlp.go (hunk 1 above)
Comment (tigrannajaryan) [reply exporter/otlphttpexporter/otlp.go:176]: Good point.

Comment (tigrannajaryan) [reply exporter/otlphttpexporter/otlp.go]: Fixed.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**